Zend/zend_operators.h

```diff
@@ -804,11 +804,17 @@ overflow: ZEND_ATTRIBUTE_COLD_LABEL
 		ZVAL_LONG(result, llresult);
 	}
 #else
-	ZVAL_LONG(result, Z_LVAL_P(op1) - Z_LVAL_P(op2));
+	/*
+	 * 'result' may alias with op1 or op2, so we need to
+	 * ensure that 'result' is not updated until after we
+	 * have read the values of op1 and op2.
+	 */
 
 	if (UNEXPECTED((Z_LVAL_P(op1) & LONG_SIGN_MASK) != (Z_LVAL_P(op2) & LONG_SIGN_MASK)
-		&& (Z_LVAL_P(op1) & LONG_SIGN_MASK) != (Z_LVAL_P(result) & LONG_SIGN_MASK))) {
+		&& (Z_LVAL_P(op1) & LONG_SIGN_MASK) != (((zend_ulong) Z_LVAL_P(op1) - (zend_ulong) Z_LVAL_P(op2)) & LONG_SIGN_MASK))) {
 		ZVAL_DOUBLE(result, (double) Z_LVAL_P(op1) - (double) Z_LVAL_P(op2));
+	} else {
+		ZVAL_LONG(result, Z_LVAL_P(op1) - Z_LVAL_P(op2));
```

Comment:
I guess we're not actually testing this branch? Probably would only be used on some pretty esoteric platforms.

Reply:
> I guess we're not actually testing this branch?

My bad (#17472) ;)


> Probably would only be used on some pretty esoteric platforms.

Possibly not even that. Would require a compiler without ASM goto support, no builtins (SADDL etc.) and non Windows system. Plus, that compiler would need to support C11 (or at least the subset we're using).

Reply:
It's better to perform the integer math once, keeping the result in a temporary variable.
Otherwise C compiler may generate longer fast path code.